rules/firestore.json

```diff
@@ -45,40 +51,132 @@ service cloud.firestore {
         }
         // joueurs will be renamed to users
         match /joueurs/{userId} {
-            // Anybody can create a user
-            allow create: if true;
-            // Only a user can update its own fields
-            allow update: if isOwner(userId);
-            // Anybody can read the fields of a user
+            // Only an authenticated account can create its own user, and they cannot directly be verified
+            allow create: if isOwner(userId) && request.resource.data.verified == false
+            // Only the user can change its fields
+            // Moreover, the username cannot be changed once set
+            // And the verified field can only be changed to true if the user has a username and has its email verified
+            allow update: if isOwner(userId) && doesNotModifyUsernameIfSet() && onlyModifiesVerifiedIfVerified();
+            // Anybody can read the fields of a user.
+            // This is because before user creation, we need to check if the username is free.
             allow read: if true;
         }
-        match /users/{userId} {
-            // Anybody can create a user
-            allow create: if true;
-            // Only a user can update its own fields
-            allow update: if isOwner(userId);
-            // Anybody can read the fields of a user
-            allow read: if true;
+        function isOwner(userId) {
+            return request != null &&
+                request.auth != null &&
+                request.auth.uid == userId;
         }
-        match /parties/{id} {
-            allow write, read: if userVerified();
+        function doesNotModifyUsernameIfSet() {
+            let usernameNotYetSet = hasField(resource.data, 'username') == false;
+            let usernameBeingSet = hasField(request.resource.data, 'username');
+            // It sometimes happens that firebase sends the username as part as updates, even though it is not actually updated. Hence, we accept usernames as part of updates, only if it doesn't change.
+            let doesNotChange = hasField(resource.data, 'username') && hasField(request.resource.data, 'username') && resource.data['username'] == request.resource.data['username'];
```

Comment:
que un seul des deux soit setté est suffisant à ce que a == b soit faux si l'autre aussi, p-e qu'on peut simplifier, sauf si tu estime que c'est mieux que ce soit encore plus explicite ?

Reply:
hm oui ici j'pense  c'est plus clair d'être explicite: `doesNotChange` veut dire : y'a une update, mais elle change rien

Reply:
par contre on a deux fois hasFiled(resource.date, 'username'); et usernameBeingSet pourrait être réutilisée, ça vaudrait la peine de créer une usernameSet et d'avoir une ligne finale qui dit
let doesNotChange = usernameSetted && usernameBeingSet && (userNameBefore == userNameAfter)

Reply:
adapté un peu plus dans ce sens, mais tu peux pas accéder `resource.data['username']` si t'es pas certain qu'il est défini, donc pas moyen de définir `usernameBefore`

Reply:
Okay je jette un oeil quand c'est pushé

Reply:
ça l'est !

Reply:
Tiens en fait:
function preservesField(field) {
            // A field is preserved if it is not part of the update
            // or if the same value is given in the update
} 
C'est pas utilisable dans ce contexte ?

Reply:
toutafévré, changé